external-crates/move/crates/move-package/tests/test_lock_file.rs

```diff
@@ -101,6 +103,71 @@ fn update_lock_file_toolchain_version() {
     expected.assert_eq(&toml);
 }
 
+#[test]
+fn test_update_managed_address() {
+    let pkg = create_test_package().unwrap();
+    let lock_path = pkg.path().join("Move.lock");
+
+    // Initialize lock file.
+    let lock = LockFile::new(
+        pkg.path().to_path_buf(),
+        /* manifest_digest */ "42".to_string(),
+        /* deps_digest */ "7".to_string(),
+    )
+    .unwrap();
+    lock.commit(&lock_path).unwrap();
+
+    // Update managed address in lock file.
+    let pb = PathBuf::from(pkg.path());
+    let mut lock = LockFile::from(pb, &lock_path).unwrap();
+    update_managed_address(
+        &mut lock,
+        "default".into(),
+        ManagedAddressUpdate::Published {
+            original_id: "0x123".into(),
+            chain_id: "35834a8a".into(),
+        },
+    )
+    .unwrap();
+
+    update_managed_address(
+        &mut lock,
+        "default".into(),
+        ManagedAddressUpdate::Upgraded {
+            latest_id: "0x456".into(),
+            version: 2,
+        },
+    )
+    .unwrap();
+    lock.commit(&lock_path).unwrap();
+
+    // Read lock file and check contents.
+    let mut lock_file = File::open(lock_path).unwrap();
+    let contents = {
+        let mut buf = String::new();
+        let _ = lock_file.read_to_string(&mut buf);
+        buf
+    };
+
+    let expected = expect![[r##"
+        # @generated by Move, please check-in and do not edit manually.
+
+        [move]
+        version = 0
+        manifest_digest = "42"
+        deps_digest = "7"
+
+        [env]
+
+        [env.default]
+        chain-id = "35834a8a"
+        original-published-id = "0x123"
+        latest-published-id = "0x456"
+        published-version = "2"
```

Comment:
Based on feedback in https://github.com/MystenLabs/sui/pull/16445#discussion_r1507766175, some notes on the new schema:

- Flatter structure under a single `env.<name>` table to capture managed address info. It's generally annoying to deal with nested structures in TOML without much of a win. To guard against other potential key names like `published-id` or `version` referring to other parts of a build or package, I've made these key names more verbose.

- Forward-compatibility: This schema doesn't prescribe how to map or design dependencies or other parts of the build and resolution wrt an environment. It simply makes the distinction that any managed address is associated with _some_ environment.
  - I have more thoughts here on how we associate environments with managed addresses and other parts of a build as per discussion: https://github.com/MystenLabs/sui/pull/16445#discussion_r1507766175. Suffice to say picking this schema shouldn't be blocking or constraining us in a meaningful way.

- Side note: I couldn't figure out how to create a `[env.default]` table without the output toml also adding an empty `[env]` table, so I'm just going to leave it at that.

Reply:
> To guard against other potential key names like published-id or version referring to other parts of a build or package, I've made these key names more verbose.

IIUC, these name conflicts would only happen because we wanted to add something new to the environment, right? In which case, that seems fine.

>Suffice to say picking this schema shouldn't be blocking or constraining us in a meaningful way. 

:+1:

> Side note: I couldn't figure out how to create a [env.default] table without the output toml also adding an empty [env] table, so I'm just going to leave it at that.

🤷 

On an unrelated note, I noticed that the Move.lock version is still 0 -- I thought we bumped this to allow for toolchain versioning to detect an old lock file that didn't know about toolchain versioning vs a new lock file that knows about toolchain versioning but hasn't seen a publish yet?

Reply:
> On an unrelated note, I noticed that the Move.lock version is still 0 -- I thought we bumped this to allow for toolchain versioning to detect an old lock file

In retrospect, yeah, this should have been bumped 😬 But! I think we can redeem the situation because toolchain info isn't used yet by default (the code using it is feature flagged). If we bump the schema at the same time when we un-feature flag, the data will become "live" from that version onward, and and we'll only regard toolchain info in version `1` or later. Not ideal for a couple of reasons, but thinking about a solution I think this works out. Will put up a PR, I think it's okay if we make toolchain versioning live now.

As for this PR, when the data is written to lock files (becomes live) I will bump the lock schema once again, but not before the version that corresponds to live toolchain version data.